docs: add page about deployment topologies (#5325)
grafana/agent: docs/sources/flow/setup/deploy-agent.md

to consider using each, issues that users may run into, as well as scaling
considerations.

## As a centralized collection service
Deploying the Grafana Agent as a centralized service is recommended for
collecting application telemetry. This allows a smaller set of agents to
coordinate service discovery, collection, and remote writing.

Using this topology requires deploying the Agent on separate infrastructure,
and making sure that agents can discover and reach these applications over the
network. The main predictor for the size of the agent is the number of active
metrics series it is scraping; a rule of thumb is ~10kb of memory for each
series. We recommend you start looking towards horizontal scaling around the 1M
active series mark.

### Using Kubernetes StatefulSets
Deploying Grafana Agent as a StatefulSet is the recommended option for metrics
collection.
The persistent pod identifiers make it possible to consistently match volumes
with pods so that you can use them for the WAL directory.

You can also use a Kubernetes deployment in cases where persistent storage is not required, such as a traces-only pipeline.

### Pros

> * Straightforward scaling using clustering or hashmod sharding

Reviewer: add links for hashmod and clustering?

Author: Done, thanks!

* Minimizes the “noisy neighbor” effect
* Easy to meta-monitor

### Cons
* Requires running on separate infrastructure

### Use for
* Scalable telemetry collection

### Don’t use for
* Host-level metrics and logs

## As a host daemon
Deploying one Grafana Agent per machine is required for collecting
machine-level metrics and logs, such as node_exporter hardware and network
metrics or journald system logs.

Each Agent requires opening an outgoing connection for each remote endpoint
it’s shipping data to. This can lead to NAT port exhaustion on the egress
infrastructure. Each egress IP can support up to (65535 - 1024 = 64511)
outgoing connections on different ports. So, if all agents are shipping metrics

> and log data, an egress IP can support up to 32,255 agents.

Author: Can someone double-check this here?

Reviewer: I don't think its an exact number because you also have any other traffic. Plus if telemetry is enabled that will consume available ports.

Author: Yeah, all we can do here is add an upper limit so that people are aware of its effect on their infrastructure.

to consider using each, issues that users may run into, as well as scaling
considerations.

## As a centralized collection service
Deploying the Grafana Agent as a centralized service is recommended for
collecting application telemetry. This allows a smaller set of agents to
coordinate service discovery, collection, and remote writing.

Using this topology requires deploying the Agent on separate infrastructure,
and making sure that agents can discover and reach these applications over the
network. The main predictor for the size of the agent is the number of active
metrics series it is scraping; a rule of thumb is ~10kb of memory for each
series. We recommend you start looking towards horizontal scaling around the 1M
active series mark.

### Using Kubernetes StatefulSets
Deploying Grafana Agent as a StatefulSet is the recommended option for metrics
collection.
The persistent pod identifiers make it possible to consistently match volumes
with pods so that you can use them for the WAL directory.

You can also use a Kubernetes deployment in cases where persistent storage is not required, such as a traces-only pipeline.

> ### Pros

Reviewer 2: looks like the heading level matches `### Using Kubernetes StatefulSets`. is this a subsection and should have 4 #s? Similar question throughout but I won't open a comment for each.

Author: It's not supposed to be a subsection of the Kubernetes deployment, but rather the idea was the following structure.

* X
* X on Kubernetes
* Pros of X
* Cons of X
* etc...

If it's not apparent I can move things around for the K8S section to be at the bottom.

Reviewer 3: I ran `make docs` on Paschalis' branch, and I think the headings are much more clear over there. So I think it's ok to just leave it as it is.

outgoing connections on different ports. So, if all agents are shipping metrics
and log data, an egress IP can support up to 32,255 agents.

### Using Kubernetes DaemonSets
The simplest use case of the host daemon topology is a Kubernetes DaemonSet,
and is required for node-level observability (for example cAdvisor metrics) and
collecting pod logs.

### Pros
* Doesn’t require running on separate infrastructure
* Typically leads to smaller-sized agents
* Lower network latency to instrumented applications

### Cons
* Requires planning a process for provisioning the Agent on new machines, as well as keeping configuration up to date to avoid configuration drift
* Not possible to scale agents independently when using Kubernetes DaemonSets
* Scaling the topology can strain external APIs (like service discovery) and network infrastructure (like firewalls, proxy servers, and egress points)

### Use for
* Collecting machine-level metrics and logs (for example node_exporter hardware metrics, Kubernetes pod logs)

### Don’t use for
* Scenarios where the Agent grows so large it can become a noisy neighbor
* Collecting an unpredictable amount of telemetry

> 

Reviewer 2: extra empty line

---
aliases:
- /docs/grafana-cloud/agent/flow/setup/deploy-agent/
- /docs/grafana-cloud/monitor-infrastructure/agent/flow/setup/deploy-agent/
- /docs/grafana-cloud/monitor-infrastructure/integrations/agent/flow/setup/deploy-agent/
canonical: https://grafana.com/docs/agent/latest/flow/setup/start-agent/
description: Learn about possible deployment topologies for Grafana Agent
menuTitle: Deploy Grafana Agent
title: Grafana Agent deployment topologies
weight: 900
---

# Deploying Grafana Agent

Grafana Agent is a flexible, vendor-neutral telemetry collector. This

> flexibility means that the Agent doesn’t enforce a specific deployment topology

Reviewer 4: ```suggestion
flexibility means that Grafana Agent doesn’t enforce a specific deployment topology
```

but can work in multiple scenarios.

This page lists common topologies used for deployments of Grafana Agent, when
to consider using each, issues that users may run into, as well as scaling
considerations.

## As a centralized collection service

> Deploying the Grafana Agent as a centralized service is recommended for

Reviewer 4: ```suggestion
Deploying Grafana Agent as a centralized service is recommended for
```

> collecting application telemetry. This allows a smaller set of agents to

Reviewer 4: ```suggestion
collecting application telemetry. This topology allows you to use a smaller number of agents to
```

Smaller number of Agents compared to what? Do we need to quantify this?

coordinate service discovery, collection, and remote writing.

> ![](../../../assets/deploy-agent/centralized-collection.png)

Reviewer 4: We have some guidelines for media that covers file name conventions and where media should be stored.

* File names: https://grafana.com/docs/writers-toolkit/write/image-guidelines/#media-asset-file-naming-conventions
* Asset upload location: https://grafana.com/docs/writers-toolkit/write/image-guidelines/#where-to-store-media-assets

This means that we should move these images out of `/docs/assets` before we merge.

Author: TIL, thanks for the heads up!

Using this topology requires deploying the Agent on separate infrastructure,
and making sure that agents can discover and reach these applications over the

> network. The main predictor for the size of the agent is the number of active

Reviewer 4: ```suggestion
network. The main predictor for the size of the agent is the number of active
```

What do you mean by "the size of the agent"?  Is this in terms of resource requirements? (RAM/CPU use?)

Author: Yes, that's what I had in mind.

> metrics series it is scraping; a rule of thumb is ~10kb of memory for each

Reviewer 4: ```suggestion
metrics series it is scraping; a rule of thumb is approximately 10 KB of memory for each
```

> series. We recommend you start looking towards horizontal scaling around the 1M

Reviewer 5: Do I understand correctly that before horizontal scaling we would recommend to run a single instance? Maybe that should be a paragraph itself. I would be cautious to recommend it though, even if it's less than 1M series, users may be recording some pretty important telemetry and may not want to have a single point of failure.

Author: > Do I understand correctly that before horizontal scaling we would recommend to run a single instance 

Not really; they can still make use of a single HA label pairs just for high-availability without properly scaling out horizontally, but I wanted to provide a rule of thumb for people to use.

> series. We recommend you start looking towards horizontal scaling around the 1M

Reviewer 4: ```suggestion
series. We recommend you start looking towards horizontal scaling around the 1 million
```

active series mark.

### Using Kubernetes StatefulSets
Deploying Grafana Agent as a StatefulSet is the recommended option for metrics
collection.
The persistent pod identifiers make it possible to consistently match volumes

> with pods so that you can use them for the WAL directory.

Reviewer 5: We all often talk about WAL, but it may be unclear to newcomers, so perhaps we can add a link to https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.remote_write/ and to Loki's equivalent? Or we could explain it here in one sentence.

Author: > If DaemonSets are good enough for logs, not sure why they wouldn't be good for metrics, at least as far as persistent storage is concerned?

Around persistent storage, the idea is that StatefulSets can mount Volumes of different [Storage Classes](https://kubernetes.io/docs/concepts/storage/storage-classes/), eg. SSDs that may be required to keep up with a higher rate of metrics, whereas for logs, the positions file is updates much less frequently and is much smaller than the WAL segments, so there's no issue using the pod's own storage.

Reviewer 3: Interesting, thank you, I didn't know that.

You can also use a Kubernetes deployment in cases where persistent storage is not required, such as a traces-only pipeline.

### Pros
* Straightforward scaling using [clustering][] or [hashmod sharding][]
* Minimizes the “noisy neighbor” effect
* Easy to meta-monitor

### Cons

> * Requires running on separate infrastructure

Reviewer 5: Nit: it's not a strict requirement I feel - you can colocate agents on the same workflows. I think that we allow for minimising the noisy-neighbour effect when agents are deployed on separate infrastructure.

### Use for
* Scalable telemetry collection

### Don’t use for
* Host-level metrics and logs

## As a host daemon
Deploying one Grafana Agent per machine is required for collecting
machine-level metrics and logs, such as node_exporter hardware and network
metrics or journald system logs.

![](../../../assets/deploy-agent/daemonset.png)

> Each Agent requires opening an outgoing connection for each remote endpoint

Reviewer 4: ```suggestion
Each Grafana Agent requires you to open an outgoing connection for each remote endpoint
```

it’s shipping data to. This can lead to NAT port exhaustion on the egress
infrastructure. Each egress IP can support up to (65535 - 1024 = 64511)
outgoing connections on different ports. So, if all agents are shipping metrics
and log data, an egress IP can support up to 32,255 agents.

### Using Kubernetes DaemonSets
The simplest use case of the host daemon topology is a Kubernetes DaemonSet,

> and is required for node-level observability (for example cAdvisor metrics) and

Reviewer 4: ```suggestion
and it is required for node-level observability (for example cAdvisor metrics) and
```

collecting pod logs.

### Pros
* Doesn’t require running on separate infrastructure
* Typically leads to smaller-sized agents
* Lower network latency to instrumented applications

> ### Cons

Reviewer 5: I'd say that typically this will also use more resources, maybe something like this:
> Large number of small agents may use more resources than fewer large agents

> * Requires planning a process for provisioning the Agent on new machines, as well as keeping configuration up to date to avoid configuration drift

Reviewer 4: ```suggestion
* Requires planning a process for provisioning Grafana Agent on new machines, as well as keeping configuration up to date to avoid configuration drift
```

* Not possible to scale agents independently when using Kubernetes DaemonSets
* Scaling the topology can strain external APIs (like service discovery) and network infrastructure (like firewalls, proxy servers, and egress points)

### Use for

> * Collecting machine-level metrics and logs (for example node_exporter hardware metrics, Kubernetes pod logs)

Reviewer 4: ```suggestion
* Collecting machine-level metrics and logs (for example, node_exporter hardware metrics, Kubernetes pod logs)
```

### Don’t use for

> * Scenarios where the Agent grows so large it can become a noisy neighbor

Reviewer 4: ```suggestion
* Scenarios where Grafana Agent grows so large it can become a noisy neighbor
```

* Collecting an unpredictable amount of telemetry

## As a container sidecar

> Deploying the Grafana Agent as a container sidecar is only recommended for

Reviewer 4: ```suggestion
Deploying Grafana Agent as a container sidecar is only recommended for
```

short-lived applications or specialized agent deployments.

![](../../../assets/deploy-agent/sidecar.png)

### Using Kubernetes pod sidecars

> In a Kubernetes environment, the sidecar model consists of deploying the agent

Reviewer 4: ```suggestion
In a Kubernetes environment, the sidecar model consists of deploying Grafana Agent
```
